### PR TITLE
rgw/reshard: fix hold reshard lock collisions

### PR DESCRIPTION
**Pull request by liangmingyuanneo**
Now, the old reshard oids retain in-progress state after resharding succeed, as a result, all operations blocked by reshard will gain reshard_lock to get new bucket info. This will lead to mass collisions, then parts of operations fail to hold lock in the loop. To solve it, update the newest bucket_obj after accqure lock failed, then redo the operations with new bucket info.

The collision easy be recurred when resharding bucket with large omap that old shards will cost a long time to clean. It also can be mocked by annotating clean_index in commit_reshard.

Signed-off-by: Mingyuan Liang <liangmingyuan@baidu.com>

### CODE REVIEW

**Comment by liangmingyuanneo** [comment]
@cbodley please review at your time

**Comment by ivancich** [comment]
@liangmingyuanneo When you get a chance, would you mind looking at #48899 and see whether it interacts with your proposed changes or is orthogonal to them. Thank you!

**Comment by liangmingyuanneo** [comment]
> @liangmingyuanneo When you get a chance, would you mind looking at #48899 and see whether it interacts with your proposed changes or is orthogonal to them. Thank you!

Thank for your attention. I think it's orthogonal. #48800 solves the problem that old shards might be recreated. But this PR focus on shards with large omap that cost a long time to be deleted after reshard complete. Many ops blocked in resharding, they try to get reshard_lock after reshard completes, but only one can get and call clear_resharding with new bucketinfo. Meanwhile, other ops get reshard_lock failed and wait for 5s. This might happy last for 10 num_retries as a result of reshard lock get collision, so this PR fetch new bucketinfo when get reshard_lock failed.

File: src/rgw/driver/rados/rgw_rados.cc [hunk 1]
```diff
@@ -6916,11 +6916,11 @@ int RGWRados::block_while_resharding(RGWRados::BucketShard *bs,
...
       if (ret < 0) {
-	ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
-	  " failed to refresh bucket info after reshard when get bucket "
-	  "resharding succeeded, error: " << cpp_strerror(-ret) << dendl;
-	return ret;
+        ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+          " failed to refresh bucket info after reshard when get bucket "
+          "resharding succeeded, error: " << cpp_strerror(-ret) << dendl;
       }
+      return ret;
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_rados.cc:6923]: can you help me understand the reasoning here? this reverts a recent change from https://github.com/ceph/ceph/pull/48899

Comment (liangmingyuanneo) [reply src/rgw/driver/rados/rgw_rados.cc:6923]: > can you help me understand the reasoning here? this reverts a recent change from #48899

I might not understand entirely, as far as I can see, there are two cases running to this place. Firstly, reshard success and the old shard has already been deleted, so stay at initial NOT_RESHARDING state. Secondly, reshard failed and the shard has changed to  NOT_RESHARDING state by clear_resharding(). No matter which case, it is better to return immediately and continue execute. It is not necessary to wait once, have I understood correctly?

File: src/rgw/driver/rados/rgw_rados.cc [hunk 2]
```diff
@@ -6949,6 +6949,16 @@ int RGWRados::block_while_resharding(RGWRados::BucketShard *bs,
...
 	  bucket_id << "; expected if resharding underway" << dendl;
+
+        // the reshard may have finished, so refresh bucket_obj to avoid
+        // acquiring reshard lock conflicts
+        ret = fetch_new_bucket_info("trying_to_refresh_bucket_info");
+        if (ret < 0) {
+          ldpp_dout(dpp, 0) << __func__ <<
+            " ERROR: failed to refresh bucket_obj for bucket " <<
+            bs->bucket.name << dendl;
+          continue; // try again
```
[the remark this answers is not in the document]
Comment (liangmingyuanneo) [reply src/rgw/driver/rados/rgw_rados.cc:6960]: > if `fetch_new_bucket_info()` returns `-ENOENT` here, we wouldn't expect a retry to help so we're better off returning the error directly. are there any other error codes from `fetch_new_bucket_info()` that _would_ be resolved by retrying?

Year, returning the error is better.

File: src/rgw/driver/rados/rgw_rados.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:6923]: thanks @liangmingyuanneo, i think you're right. if `fetch_new_bucket_info()` succeeds, we don't know whether that new bucket instance is resharding, so it's best to return directly to `guard_reshard()` to retry the operation. if we find that it is still resharding, we'll call back into `block_while_resharding()`

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2023-03-13_18:22:40-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**